src/memo/memo.ml

```diff
@@ -1233,101 +1240,105 @@ end = struct
       | true -> Cached_value.create value ~deps_rev
       | false -> Cached_value.confirm_old_value ~deps_rev old_cv)
 
-  and newly_considering :
-        'i 'o. ('i, 'o) Dep_node.t -> 'o Cached_value.t Sample_attempt.t =
+  and start_considering : 'i 'o. ('i, 'o) Dep_node.t -> 'o Considering_state.t =
    fun dep_node ->
     if !Counters.enabled then incr Counters.nodes_considered;
-    let dag_node = create_dag_node dep_node.without_state in
-    let restore_from_cache_frame =
-      Stack_frame_with_state.create Restore_from_cache ~dag_node
-        ~dep_node:dep_node.without_state
-    in
     let restore_from_cache =
-      Once.create ~must_not_raise:(fun () ->
-          Call_stack.push_frame restore_from_cache_frame (fun () ->
-              let+ restore_result =
-                restore_from_cache dep_node.last_cached_value
-              in
-              (match restore_result with
-              | Ok cached_value ->
-                dep_node.state <- Not_considering;
-                if !Counters.enabled then
-                  Counters.edges_considered :=
-                    !Counters.edges_considered + Deps.length cached_value.deps
-              | Failure _ -> ());
-              restore_result))
+      Computation.create ~phase:Restore_from_cache
+        ~dep_node:dep_node.without_state (fun _stack_frame ->
+          let+ restore_result = restore_from_cache dep_node.last_cached_value in
+          (match restore_result with
+          | Ok cached_value ->
+            dep_node.state <- Not_considering;
+            if !Counters.enabled then
+              Counters.edges_considered :=
+                !Counters.edges_considered + Deps.length cached_value.deps
+          | Failure _ -> ());
+          restore_result)
     in
+    (* CR-someday amokhov: When doing an incremental zero rebuild, we don't need
```

Comment:
I wish we just made the representation of non-forced `Computation` cheap. There's no reason for it to be more expensive than the `lazy` we'd be replacing it with. (the both just need to allocate a tuple with the info that will be needed in the future)

Reply:
Agreed, I've added this remark to the comment.